File: java2/src/구명보트/구명보트.java
package 구명보트;
import java.util.Arrays;

public class 구명보트 {
    public static int solution(int[] people, int limit) {
        Arrays.sort(people);

        int start = 0;
        int end = people.length - 1;

        int rescue = 0;
        while (start <= end) {
            if (people[start] + people[end] > limit) {
                rescue++;
                end--;
            } else {
                rescue++;
                end--;
                start++;
            }
        }

        return rescue;
    }


    public static void main(String[] args) {
//        int[] people = {70, 50, 80, 50};
//        int limit = 100;
//        System.out.println(solution(people, limit));
        // 3

        int[] people2 = {70, 80, 50};
        int limit2 = 100;
        System.out.println(solution(people2, limit2));
        // 3

    }
}


// https://school.programmers.co.kr/learn/courses/30/lessons/42885?language=java

File: python/algorithm/everything_algorithm/algorithm_study/programmers/level2/구명보트/구명보트.py
def solution(people, limit):
    answer = 0

    p_sort = sorted(people)
    print(p_sort)

    sum = 0
    arr = []
    for key, el in enumerate(p_sort):
        print('key : ', key)
        print('el : ', el)
        sum += el
        if sum == limit:
            arr.append(sum)
            sum = 0
        elif sum >= limit:
            sum -= el
            arr.append(p_sort[key - 1])
            arr.append(p_sort[key])
            sum = 0
    print('arr : ', arr)

    return answer


# print(solution([70, 50, 80, 50], 100)) # 3
print(solution([70, 80, 50], 100)) # 3

# https://school.programmers.co.kr/learn/courses/30/lessons/42885